### PR TITLE
MULE-9855 Make MuleMessage immutable. Step 1

### PR DESCRIPTION
**Pull request by dfeist**
PENDING BUILD.

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java [hunk 1]
```diff
@@ -1,86 +0,0 @@
...
- */
-package org.mule.runtime.core;
-
-import static org.junit.Assert.assertEquals;
-
-import org.mule.runtime.core.api.MuleMessage;
-import org.mule.runtime.core.api.ThreadSafeAccess;
-import org.mule.tck.junit4.AbstractMuleContextTestCase;
-
-public abstract class AbstractThreadSafeAccessTestCase extends AbstractMuleContextTestCase
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java]: Would it make sense to keep testing this for events?

File: core-tests/src/test/java/org/mule/runtime/core/mule/enricher/MessageEnricherTestCase.java [hunk 2]
```diff
@@ -344,7 +343,7 @@ public void enricherConservesSameEventInstanceNonBlockingTargetNonBlocking() thr
 
         nullReplyToHandler.latch.await(RECEIVE_TIMEOUT, TimeUnit.MILLISECONDS);
 
-        assertThat(sensingNullMessageProcessor.event.getMessage(), not(sameInstance(in.getMessage())));
+        assertThat(sensingNullMessageProcessor.event.getMessage(), sameInstance(in.getMessage()));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/mule/enricher/MessageEnricherTestCase.java]: is this assertion still meaningful?

File: core/src/main/java/org/mule/runtime/core/el/context/OutboundAttachmentMapContext.java [hunk 3]
```diff
@@ -49,7 +50,7 @@ public void doRemove(String key)
     @Override
     public void clear()
     {
-        event.setMessage(MuleMessage.builder(event.getMessage()).clearOutboundAttachments().build());
+        event.setMessage(MuleMessage.builder(event.getMessage()).outboundAttachments(emptyMap()).build());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/el/context/OutboundAttachmentMapContext.java]: did we remove the clear... method from the builder?

File: transports/vm/src/main/java/org/mule/compatibility/transport/vm/VMMessageReceiver.java [hunk 4]
```diff
@@ -115,10 +113,6 @@ public MuleEvent process() throws Exception
                     if (event != null && !VoidMuleEvent.getInstance().equals(event) && getEndpoint().getExchangePattern().hasResponse())
                     {
                         MuleMessage returnedMessage = event.getMessage();
```
Comment (elrodro83) [inline transports/vm/src/main/java/org/mule/compatibility/transport/vm/VMMessageReceiver.java]: remove this line too

**Comment by elrodro83** [comment]
Is there any way to also check there are no synchronization on a message object?

File: core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java (hunk 1 above)
Comment (dfeist) [inline core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java]: Event doesn't use thread safe access.

File: core-tests/src/test/java/org/mule/runtime/core/mule/enricher/MessageEnricherTestCase.java (hunk 2 above)
Comment (dfeist) [inline core-tests/src/test/java/org/mule/runtime/core/mule/enricher/MessageEnricherTestCase.java]: Less meaningful, but no harm asserting that message isn't copied.  Can remove if you prefer..

File: core/src/main/java/org/mule/runtime/core/el/context/OutboundAttachmentMapContext.java (hunk 3 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/el/context/OutboundAttachmentMapContext.java]: Yes.  Do you think we should put it back?

File: core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java (hunk 1 above)
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/AbstractThreadSafeAccessTestCase.java]: It was delegating to its message

**Pull request merged**